converter.go

```diff
@@ -762,3 +763,306 @@ func higherPrecisionEnabled(ctx context.Context) bool {
+			tb := array.NewTimestampBuilder(pool, &arrow.TimestampType{TimeZone: "UTC"})
+			if col.DataType().ID() == arrow.STRUCT {
+				structData := array.NewStructData(data)
+				epoch := array.NewInt64Data(structData.Field(0).Data()).Int64Values()
+				fraction := array.NewInt32Data(structData.Field(1).Data()).Int32Values()
+				for i := 0; i < int(numRows); i++ {
+					if !col.IsNull(i) {
+						val := time.Unix(epoch[i], int64(fraction[i]))
+						tb.Append(arrow.Timestamp(val.UnixNano()))
+					} else {
+						tb.AppendNull()
+					}
+				}
+			} else {
+				for i, t := range array.NewInt64Data(data).Int64Values() {
+					if !col.IsNull(i) {
+						q := t / int64(math.Pow10(int(srcColumnMeta.Scale)))
+						r := t % int64(math.Pow10(int(srcColumnMeta.Scale)))
+						val := time.Unix(q, r)
+						tb.Append(arrow.Timestamp(val.UnixNano()))
+					} else {
+						tb.AppendNull()
+					}
+				}
+			}
+			converted = true
+			cols = append(cols, tb.NewArray())
+		case timestampTzType:
+			tb := array.NewTimestampBuilder(pool, &arrow.TimestampType{})
+			structData := array.NewStructData(data)
+			if structData.NumField() == 2 {
+				epoch := array.NewInt64Data(structData.Field(0).Data()).Int64Values()
+				timezone := array.NewInt32Data(structData.Field(1).Data()).Int32Values()
+				for i := 0; i < int(numRows); i++ {
+					if !col.IsNull(i) {
+						loc := Location(int(timezone[i]) - 1440)
+						tt := time.Unix(epoch[i], 0)
+						val := tt.In(loc)
+						tb.Append(arrow.Timestamp(val.UnixNano()))
+					} else {
+						tb.AppendNull()
+					}
+				}
+			} else {
+				epoch := array.NewInt64Data(structData.Field(0).Data()).Int64Values()
+				fraction := array.NewInt32Data(structData.Field(1).Data()).Int32Values()
+				timezone := array.NewInt32Data(structData.Field(2).Data()).Int32Values()
+				for i := 0; i < int(numRows); i++ {
+					if !col.IsNull(i) {
+						loc := Location(int(timezone[i]) - 1440)
+						tt := time.Unix(epoch[i], int64(fraction[i]))
+						val := tt.In(loc)
+						tb.Append(arrow.Timestamp(val.UnixNano()))
+					} else {
+						tb.AppendNull()
+					}
+				}
+			}
+			converted = true
+			cols = append(cols, tb.NewArray())
+		}
+		// append raw column if there is no need to transform it
+		if !converted {
+			cols = append(cols, col)
+		}
+	}
+	return array.NewRecord(s, cols, numRows), nil
+}
+
+func recordToSchema(sc *arrow.Schema, rowType []execResponseRowType) (*arrow.Schema, error) {
+	var fields []arrow.Field
+	for i := 0; i < len(sc.Fields()); i++ {
+		f := sc.Field(i)
+		srcColumnMeta := rowType[i]
+		converted := false
+
+		switch getSnowflakeType(strings.ToUpper(srcColumnMeta.Type)) {
+		case fixedType:
+			switch f.Type.ID() {
+			case arrow.DECIMAL:
+				if srcColumnMeta.Scale == 0 {
+					newField := arrow.Field{
+						Name:     f.Name,
+						Type:     &arrow.Int64Type{},
+						Nullable: f.Nullable,
+						Metadata: f.Metadata,
+					}
+					converted = true
+					fields = append(fields, newField)
```

Comment:
why don't you have `newField = f` at the beginning and one append call at the end of the `case`